### PR TITLE
Calculate transient memory usage for load of attribute vector.

### PR DESCRIPTION
**Pull request by toregge**
@geirst : PTAL.

### CODE REVIEW

File: searchcore/src/vespa/searchcore/proton/attribute/attribute_usage_sampler_functor.cpp [hunk 1]
```diff
@@ -2,10 +2,49 @@
...
+                ;
+            }
+        }
+    }
+    return 0u;
+}
+
+}
+
 AttributeUsageSamplerFunctor::AttributeUsageSamplerFunctor(
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/attribute/attribute_usage_sampler_functor.cpp:48]: Also consider adding unit test for this class.

File: searchcore/src/vespa/searchcore/proton/server/resource_usage_explorer.cpp [hunk 2]
```diff
@@ -47,6 +47,8 @@ ResourceUsageExplorer::get_state(const vespalib::slime::Inserter &inserter, bool
         memory.setDouble("utilization", usageState.memoryState().utilization());
         memory.setLong("physicalMemory", _usageFilter.getHwInfo().memory().sizeBytes());
         convertMemoryStatsToSlime(_usageFilter.getMemoryStats(), memory.setObject("stats"));
+        size_t transient_memory = _usageFilter.get_transient_memory_usage();
+        memory.setLong("transient", transient_memory);
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/resource_usage_explorer.cpp:51]: Consider adding metric that exposes the transient memory usage in ResourceUsageMetrics.

**Review by geirst — COMMENTED** [review]
Overall looks good, some comments to consider.

File: searchcore/src/vespa/searchcore/proton/attribute/attribute_usage_sampler_functor.cpp [hunk 3]
```diff
@@ -2,10 +2,49 @@
...
+#include <vespa/searchlib/attribute/loadedvalue.h>
+
+using search::attribute::BasicType;
 
 namespace proton {
 
+namespace {
+
+size_t
+get_transient_memory_usage(const search::attribute::Config& old_config,
```
[the remark this answers is not in the document]
Comment (toregge) [reply searchcore/src/vespa/searchcore/proton/attribute/attribute_usage_sampler_functor.cpp]: Fixed.

File: searchcore/src/vespa/searchcore/proton/common/i_transient_memory_usage_provider.h [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+#pragma once
+
+#include <cstddef>
+
+namespace proton {
+
+/*
+ * Interface class providing transient memory usage, e.g. extra memory needed
+ * for loading or saving an attribute vector.
```
[the remark this answers is not in the document]
Comment (toregge) [reply searchcore/src/vespa/searchcore/proton/common/i_transient_memory_usage_provider.h]: Class comment has been updated.

File: searchcore/src/vespa/searchcore/proton/attribute/attribute_usage_sampler_context.h [hunk 5]
```diff
@@ -8,6 +8,8 @@
 namespace proton {
 
 class AttributeUsageFilter;
+class AttributeConfigInspector;
+class TransientMemoryUsageProvider;
 
 /*
  * Context for sampling attribute usage stats.  When instance is
```
[the remark this answers is not in the document]
Comment (toregge) [reply searchcore/src/vespa/searchcore/proton/attribute/attribute_usage_sampler_context.h]: Fixed.

File: searchcore/src/vespa/searchcore/proton/attribute/attribute_usage_sampler_functor.cpp (hunk 1 above)
Comment (toregge) [reply searchcore/src/vespa/searchcore/proton/attribute/attribute_usage_sampler_functor.cpp:13]: Done.

File: searchcore/src/vespa/searchcore/proton/server/resource_usage_explorer.cpp (hunk 2 above)
Comment (toregge) [reply searchcore/src/vespa/searchcore/proton/server/resource_usage_explorer.cpp:51]: Done

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**